### PR TITLE
Added a new module, llm_training because that's what we all should be doint right now

### PR DESCRIPTION
**Pull request by Fabinout**
I didn't generate a gif, but here is a transcript of a training 

> 
> [INFO] Device found: NVIDIA GeForce RTX 3080 Ti (12288MB)
> [INFO] Using mixed precision training (fp16).
> [INFO] Gradient accumulation steps: 16 | Total batch size: 512
> [INFO] Initializing model weights from 'qwen2.5-7b'... Done.
> 100%|[██████████]| 547.93KiB/547.93KiB [00:00<00:00, 34.64MiB/s]
> 
> 100%|[██████████]| 61.36KiB/61.36KiB [00:00<00:00, 75.57MiB/s]
> 
> 100%|[██████████]| 3.19GiB/3.19GiB [02:53<00:00, 18.75MiB/s]
> 
> 100%|[██████████]| 7.33KiB/7.33KiB [00:00<00:00, 9.22MiB/s]
> 
> Training set has 18568 instances
> 

In this state, i like what it outputs, but if you have any insights or input i can polish it a bit more

Implements #690.

### CODE REVIEW

**Comment by Fabinout** [comment]
Pfff i did it quickly while cooking lunch for the kids. That's shameful, i'll look into it tonight.

File: src/modules/llm_training.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        let num_epochs: u32 = rng.random_range(3..8);
+        let steps_per_epoch: u32 = rng.random_range(500..2000);
+        let mut loss: f64 = rng.random_range(3.0..5.0);
+
+        let epoch_bar_steps = 12u32;
+        let epoch_step_ms = 80u64;
+
+        let simulated_epoch_secs: u64 = rng.random_range(60..300);
+
+        for epoch in 1..=num_epochs {
```
Comment (svenstaro) [inline src/modules/llm_training.rs]: Add a comment for what this whole for loop is for.

**Review by svenstaro — CHANGES_REQUESTED** [review]

File: src/modules/llm_training.rs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+pub struct Gpu<'a> {
+    pub name: &'a str,
+    pub vram: u64,
+}
+
+pub struct LlmTraining;
+
+fn fmt_time(secs: u64) -> String {
+    format!("{:02}:{:02}", secs / 60, secs % 60)
+}
```
[the remark this answers is not in the document]
Comment (Fabinout) [reply src/modules/llm_training.rs]: it converts a number of seconds in a time that reads "05:45', that's used in several lines of output.
i didn't write format in plain word because it felt like it wasn't how it was done in rust (like the command cargo fmt).
Should I rename it format_mm_ss for exemple ?

File: src/modules/llm_training.rs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            for i in 0..=epoch_bar_steps {
+                erase_line().await;
+
+                let step = (steps_per_epoch * i / epoch_bar_steps).min(steps_per_epoch);
+                let pct = step as f64 / steps_per_epoch as f64 * 100.0;
+                let step_loss =
+                    loss - rng.random_range(0.0_f64..0.2) * i as f64 / epoch_bar_steps as f64;
+                epoch_bar.replace(step as usize);
+
+                if i == 0 {
```
[the remark this answers is not in the document]
Comment (svenstaro) [reply src/modules/llm_training.rs]: No, I'd like the code to be obvious for contributors when they want to make changes.

File: src/data.rs [hunk 4]
```diff
@@ -15,6 +15,8 @@ static RKHUNTER_CHECKS: &str = include_str!("../data/rkhunter_checks.txt");
 static RKHUNTER_ROOTKITS: &str = include_str!("../data/rkhunter_rootkits.txt");
 static RKHUNTER_TASKS: &str = include_str!("../data/rkhunter_tasks.txt");
 static JULIA_PACKAGES: &str = include_str!("../data/julia.csv");
+static LLM_GPUS: &str = include_str!("../data/llm_gpus.csv");
```
Comment (svenstaro) [inline src/data.rs]: Why is this a csv? It's not a csv.

File: data/llm_models.txt [hunk 5]
```diff
@@ -0,0 +1,23 @@
+GLM-5
+Kimi-K2.5
+K2-Think-V2
+Qwen3.5-27b
```
Comment (svenstaro) [inline data/llm_models.txt]: The file is called `llm_models.txt` but LLM already means "Large Language Model". So this is "large_language_model_models.txt" which doesn't really make sense.

**Review by svenstaro — CHANGES_REQUESTED** [review]

**Comment by Fabinout** [comment]
Looks like there is still a lot of issues, i'll check it again next week. Sorry for the inconveniance !
I'm sorry I wasted your time, i'll try to check it twice before i ask for a review again.

**Comment by svenstaro** [comment]
Hey, do you wanna try to finish this up?

**Comment by Fabinout** [comment]
Sure, but i don't want to waste your time even more 😅.
I've checked and the code was ok but i want to make sure i wrote enough comments.
Tomorrow evening !

**Comment by Fabinout** [comment]
I think this is ready for review @svenstaro, hope you'll like it !

**Comment by svenstaro** [comment]
Please rebase, there's a conflict.

**Comment by Fabinout** [comment]
Will do right now, and sqsuash my commits

File: src/modules/llm_training.rs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+                epoch_bar_steps,
+                epoch_step_ms,
+                simulated_epoch_secs,
+                |i, step, pct, elapsed_secs, remaining_secs, bar| {
+                    if i == 0 {
+                        format!(
+                            "Epoch {epoch}/{num_epochs}:   0%[          ] 0/{steps_per_epoch} [00:00<?, ?it/s]"
+                        )
+                    } else {
+                        let mut local_rng = rand::rng();
```
Comment (svenstaro) [inline src/modules/llm_training.rs]: Why are you creating a new rng here?

File: src/modules/llm_training.rs [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+            if appconfig.should_exit() {
+                return;
+            }
+        }
+
+        let val_steps: u32 = rng.random_range(50..300);
+        let val_simulated_secs: u64 = rng.random_range(3..15);
+        let val_bar_steps = 12u32;
+        let val_step_ms = 80u64;
```
Comment (svenstaro) [inline src/modules/llm_training.rs:246]: Do you really need all the explicit type annotations here and everywhere else? Try to remove them in all places that are possible to improve readability.

**Review by svenstaro — CHANGES_REQUESTED** [review]

**Comment by svenstaro** [comment]
By the way, if you want to, you can spice up the output a bit using `yansi` (see other modules. I'm not sure what exact command you're modeling this after, though. Maybe what you're trying to model doesn't have color output, either. :)

At any rate, I think we're almost ready to merge. Also remember to update the README with the new module.

File: src/modules/llm_training.rs [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    fn signature(&self) -> String {
+        "python train.py --model gpt2 --dataset wikitext".to_string()
+    }
+
+    async fn run(&self, appconfig: &AppConfig) {
+        let mut rng = rng();
+
+        // selects a random GPU and logs the device details
+        let gpu = LLM_GPUS_LIST.choose(&mut rng).unwrap();
+        let (gpu_name, gpu_vram) = (gpu.name, gpu.vram);
```
Comment (svenstaro) [inline src/modules/llm_training.rs]: I think this doesn't really make sense. You save no characters by doing this destructuring. I think you can just use `gpu.name` and `gpu.vram` directly where needed.

File: src/modules/llm_training.rs [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+            let speed: u64 = rng.random_range(1_000_000..80_000_000);
+            let simulated_total_secs = (file_size / speed).max(1);
+
+            let size_opts = FormatSizeOptions::from(humansize::BINARY).space_after_value(false);
+            let speed_opts = FormatSizeOptions::from(humansize::BINARY)
+                .space_after_value(false)
+                .suffix("/s");
+
+            run_progress_bar(
+                file_size as usize,
```
Comment (svenstaro) [inline src/modules/llm_training.rs:145]: This will clip on wasm because wasm's usize is only 32bit but you generate a random file size that is potentially larger than this.

File: src/modules/llm_training.rs (hunk 7 above)
Comment (Fabinout) [reply src/modules/llm_training.rs:246]: i'm too much used to java :D

File: src/modules/llm_training.rs (hunk 6 above)
Comment (Fabinout) [reply src/modules/llm_training.rs]: that's because before a refactoring there was a closure here and i thought i needed a mutable reference here.¯\_(ツ)_/¯ but instantiating a new rand every frame is definitely a mistake here. I fixed it.

File: src/modules/llm_training.rs (hunk 9 above)
Comment (Fabinout) [reply src/modules/llm_training.rs:145]: snap, i never even tried to run it on wasm you're right.
I could have a different value for web assembly, or I can limit the size of the file to 32 bit, or change the run_progress_bar implementation. What would you suggest?

**Comment by Fabinout** [comment]
I agree the output seems very bland compared to the other modules right now, i thought i would merge it, and then work on making it more colorful although i'm colorblind.

**Comment by Fabinout** [comment]
@svenstaro did you have time to check if the modifications suit you ?
I have a question for web assembly, i don't know how i should fix it

File: src/modules/llm_training.rs (hunk 9 above)
Comment (svenstaro) [reply src/modules/llm_training.rs:145]: Well you could just switch the unit to kilobytes. It's not like we care about bytes anyway. :)
